core/src/main/kotlin/gateway/MasterGateway.kt

```diff
@@ -32,15 +34,19 @@ class MasterGateway(
 
     @OptIn(FlowPreview::class)
     val events: Flow<ShardEvent> = gateways.entries.asFlow()
-            .map { (shard, gateway) -> gateway.events.map { ShardEvent(it, gateway, shard) } }
-            .flattenMerge(gateways.size.coerceAtLeast(1))
+        .map { (shard, gateway) -> gateway.events.map { ShardEvent(it, gateway, shard) } }
+        .flattenMerge(gateways.size.coerceAtLeast(1))
 
     /**
      * Calls [Gateway.start] on each Gateway in [gateways], changing the [GatewayConfiguration.shard] for each Gateway.
      */
-    suspend fun start(configuration: GatewayConfiguration) = gateways.entries.forEach { (shard, gateway) ->
-        val config = configuration.copy(shard = configuration.shard.copy(index = shard))
-        gateway.start(config)
+    suspend fun start(configuration: GatewayConfiguration): Unit = coroutineScope {
```

Comment:
launch and join jobs so the starting blocks instead, or we will end up exiting main after login happens.

```kt
    suspend fun start(configuration: GatewayConfiguration) = coroutineScope {
        val jobs = gateways.entries.map { (shard, gateway) ->
            val config = configuration.copy(shard = configuration.shard.copy(index = shard))
            launch { gateway.start(config) }
        }
        jobs.onEach { it.join() }
    }
    ```

Reply:
If you take a look at the docs of coroutineScope docs they say
> This function returns as soon as the given block and **all its children coroutines are completed**
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html

as all gateway coroutines run as long the connection is alive coroutineScope should suspend as long as any gateway connection is alive
meaning it doesn't exit main and also did not when trying this locally

Reply:
Hmm good point, this would be redundant